### DISCUSSION TITLE
Config help - Excluding a package from a group in specific cases

### DISCUSSION THREAD

**Discussion by lmilbaum** [renovatebot/renovate#26113] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, select which platform you are using.

github.com

### What is your question?

I'd like to exclude the package docker.io/golang from the dockerfile group when major and minor updates are available. My config is:
```
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "github>platform-engineering-org/.github"
  ],
  "packageRules": [
    {
      "matchPackagePatterns": ["docker.io/golang"],
      "matchDepTypes": ["major", "minor"],
      "dependencyDashboardApproval": true,
      "groupName": null
    },
    {
       "description": "Group dockerfile dependencies",
       "matchManagers": [ "dockerfile" ],
       "groupName": "dockerfile"
    }
  ]
}
```
The result is all docker dependencies are grouped in one PR including docker.io/golang even if the version bump is a minor update.

### Logs (if relevant)

<details><summary>Logs</summary>

```

 INFO: Ignoring upgrade collision (repository=lmilbaum/hyperconverged-cluster-operator, packageFile=build/Dockerfile, branch=renovate/dockerfile)
       "manager": "dockerfile",
       "depName": "docker.io/golang",
       "currentValue": "1.20.10",
       "previousNewValue": "1.21.4",
       "thisNewValue": "1.20.10"

```

</details>

**Comment by viceice** [comment]
change package rules order, they are applied from top to bottom, so last rule wins.

see tip 
https://docs.renovatebot.com/configuration-options/#packagerules

**Reply by lmilbaum** [reply]
[Changed the order](https://github.com/lmilbaum/hyperconverged-cluster-operator/blob/main/renovate.json).  Similar results.  See [Dependency Dashboard.](https://github.com/lmilbaum/hyperconverged-cluster-operator/issues/8)

**Reply by lmilbaum** [reply]
Trying a different approach. I think this one working:
```

    {
       "description": "Group dockerfile dependencies",
       "matchManagers": [ "dockerfile" ],
       "excludePackageNames": ["docker.io/golang"],
       "groupName": "dockerfile"
    },
    {
      "matchPackagePatterns": ["docker.io/golang"],
      "matchDepTypes": ["patch", "digest"],
      "groupName": "dockerfile"
    },
    {
      "matchPackagePatterns": ["docker.io/golang"],
      "matchDepTypes": ["major", "minor"],
      "dependencyDashboardApproval": true,
      "groupName": null
    }
```

**Reply by lmilbaum** [reply]
I was expecting this option to generate two PRs. One for dockerfile group, where the docker.io/golang package gets digested while another PR for bumping to a new minor release. The result is only one dockerfile PR.

**Reply by viceice** [reply]
https://docs.renovatebot.com/configuration-options/#separateminorpatch ?

**Reply by lmilbaum** [reply]
Thanks for referring to this configuration key. I didn't find the one which separates minor and digest.

**Reply by lmilbaum** [reply]
In the meanwhile, I've updated the configuration. Wasn't sure which package Rule should have the key so I added it to both of them:
```

    {
       "description": "Group dockerfile dependencies",
       "matchManagers": [ "dockerfile" ],
       "excludePackageNames": ["docker.io/golang"],
       "groupName": "dockerfile"
    },
    {
      "matchPackagePatterns": ["docker.io/golang"],
      "matchDepTypes": ["patch", "digest"],
      "separateMinorPatch": true,
      "groupName": "dockerfile"
    },
    {
      "matchPackagePatterns": ["docker.io/golang"],
      "matchDepTypes": ["major", "minor"],
      "separateMinorPatch": true,
      "dependencyDashboardApproval": true,
      "groupName": null
    }
```

**Reply by lmilbaum** [reply]
Adding the seperateMinorPatch key has not changed the result

**Reply by lmilbaum** [reply]
But the package update landed in a the all group PR ignoring the configuration completely

**Reply by lmilbaum** [reply]
Something is totally off with how renovate is working. Maybe my configuration is wrong.

**Reply by viceice** [reply]
the separate options can't be used with match update types.

**Reply by lmilbaum** [reply]
Would you mind clarifying how it could be used? I'm totally lost at the moment. Tried 10x if not more options so far.

**Reply by viceice** [reply]
```json
{
   "description": "Group dockerfile dependencies",
   "matchManagers": [ "dockerfile" ],
   "excludePackageNames": ["docker.io/golang"],
   "groupName": "dockerfile"
},
{
  "matchPackagePatterns": ["docker.io/golang"],
  "separateMinorPatch": true,
},
{
  "matchPackagePatterns": ["docker.io/golang"],
  "matchDepTypes": ["major", "minor"],
  "dependencyDashboardApproval": true,
  "groupName": null
}
```

`separateMinorPatch` can't be used together with `matchDepTypes` because the update type isn't known when `separateMinorPatch` is applied. it's applied before deciding the update type.

**Reply by lmilbaum** [reply]
Thanks @viceice for looking into it
